### PR TITLE
Feature/global refactoring

### PR DESCRIPTION
**Pull request by Pzixel**
This PR doesn't contain any logic changes for easier review. I have found some places with possible memory leaks and decorated them with TODO.

It contains split of `core.rs` in several separated files (such as `mat.rs`), which is big enough, but it may be fixed with removing `frombytes` trait (could be replaced with native `at` function call), as well as breaking down of `opencv-wrapper` in multiple little files. Looks much nicer to me:

image

There is also `gpu` -> `cuda` rename to be more spesific as well as because OpenCV may also provide support for OpenCL, TensorFlow or something else in future.

### CODE REVIEW

File: src/highgui.rs [hunk 1]
```diff
@@ -114,3 +117,20 @@ pub enum MouseEventType {
     /// Positive/negative means right and left scrolling.
     MouseHWheel = 11,
 }
+
+/// Provides some highgui functionallity
+pub trait Shower {
```
Comment (Pzixel) [inline src/highgui.rs]: Is not the best name

Comment (nebgnahz) [reply src/highgui.rs]: Verb may be better, like `Show`.
Or just `HighGui` or `Gui`?

File: native/cuda/main.cc [hunk 2]
```diff
@@ -1,10 +1,9 @@
 #include <opencv2/cudaobjdetect.hpp>
 
-#include "opencv-gpu.h"
-#include "opencv-wrapper.h"
-#include "utils.h"
+#include "../utils.h"
+#include "main.h"
```
Comment (nebgnahz) [inline native/cuda/main.cc]: main suggests that there might be a main function and this is a binary executable.

we could do `native/cuda/cuda.[h|cc]`

File: native/cuda/main.h [hunk 3]
```diff
@@ -68,7 +65,6 @@ int cv_gpu_cascade_get_min_neighbors(cv::Ptr<cv::cuda::CascadeClassifier>*);
 Size2i cv_gpu_cascade_get_max_object_size(cv::Ptr<cv::cuda::CascadeClassifier>*);
 Size2i cv_gpu_cascade_get_min_object_size(cv::Ptr<cv::cuda::CascadeClassifier>*);
```
Comment (nebgnahz) [inline native/cuda/main.h]: Mmm, since we are renaming everything to cuda, what about changing these to `cv_cuda_*`?

**Comment by nebgnahz** [comment]
I mainly have some naming suggestions.

**Pull request merged**